### PR TITLE
server: drop read request if the leader is changed

### PR DESCRIPTION
**Pull request by nolouch**
The read request will block the following read request until the context timeout if meet network partition, and this PR will drop the read request if the leader is changed. ref:  https://github.com/etcd-io/etcd/pull/9567 https://github.com/etcd-io/etcd/issues/9566

### CODE REVIEW

File: etcdserver/server.go [hunk 1]
```diff
@@ -938,7 +941,9 @@ func (s *EtcdServer) run() {
 					s.compactor.Resume()
 				}
 			}
-
+			if newLeader {
+				s.leaderChanged <- struct{}{}
```
Comment (gyuho) [inline etcdserver/server.go]: Handle when `linearizableReadLoop` had returned from a stopped server?

```go
select {
case s.leaderChanged <- struct{}{}:
default:
}
```

File: etcdserver/server.go [hunk 2]
```diff
@@ -938,7 +941,12 @@ func (s *EtcdServer) run() {
 					s.compactor.Resume()
 				}
 			}
-
+			if newLeader {
+				select {
+				case s.leaderChanged <- struct{}{}:
```
Comment (xiang90) [inline etcdserver/server.go]: if we want to do a broadcast style notification, we should close this channel not just sending one item.

Comment (nolouch) [reply etcdserver/server.go]: What do you mean?  seems I do not need to broadcast the leader change, it just consumes the channel in `linearizableReadLoop`. There are two possibilities of being blocked in the loop, one is no read request before the leader changed, and another is that there is a read request coming during the leader changed process. Although there are two places to consume this channel, it only consumes once. cc @xiang90

Comment (xiang90) [reply etcdserver/server.go]: The chan is named as `leaderChanged`. By reading the variable name, people will assume they will get notification when there is a leader change by using this chan. However, when this chan gets used by multiple routines, this chan only deliver one notification. There will be a problem.

I would rather make this chan a broadcast chan.

**Comment by wenjiaswe** [comment]
@wenjiaswe

**Comment by xiang90** [comment]
The general change looks good to me. defer to @gyuho after fixing the CI issue.

**Comment by gyuho** [comment]
Thanks @xiang90. Will take a look at CIs this week.

**Comment by xiang90** [comment]
@gyuho can you check the CI when you have time? thanks!

**Comment by gyuho** [comment]
@xiang90 There were some flaky tests from clientv3. Not related. Merging on greens. Thanks!

**Pull request merged**